tensorflow/python/keras/benchmarks/keras_examples_benchmarks/mnist_conv_custom_training_benchmark_test.py

```diff
@@ -0,0 +1,320 @@
+                          distribution_strategy=None):
+    """Run models and measure the performance.
+
+    Arguments:
+      model_fn: Model function to be benchmarked.
+      dataset: `tf.data` dataset. Should return a tuple
+        of either (inputs, targets) or (inputs, targets, sample_weights).
+      loss_fn: `tf.keras.losses.Loss` instance.
+      optimizer: `tf.keras.optimizers` instance.
+      batch_size: Integer. Number of samples per gradient update.
+      If unspecified, `batch_size` will default to 32.
+      run_iters: Integer. Number of iterations to run the performance
+        measurement. If unspecified, `run_iters` will default to 4.
+      epochs: Integer. Number of epochs to train the model.
+        If unspecified, `epochs` will default to 2.
+      distribution_strategy: Distribution strategies. It could be
+      `multi_worker_mirrored`, `one_device`, `mirrored`. If unspecified,
+      `distribution_strategy` will default to 'off'. Note that, `TPU`
+      and `parameter_server` are not supported yet.
+
+    Returns:
+      Performance summary, which contains build_time, avg_epoch_time,
+      wall_time, exp_per_sec, epochs, warmup_time, train_step_time.
+
+    Raise:
+      ValueError: if `dataset` is None or if `optimizer` instance is
+      not provided or if `loss_fn` instance is not provided.
+    """
+    if distribution_strategy is not None and \
+      not isinstance(dataset, tf.distribute.DistributedDataset):
+      raise ValueError('tf.distribute.DistributedDataset'
+                       ' required in distribution strategy.')
+
+    if distribution_strategy is None and \
+      not isinstance(dataset, tf.data.Dataset):
+      raise ValueError('`tf.data` is required.')
+
+    if not isinstance(loss_fn, tf.keras.losses.Loss):
+      raise ValueError('`tf.keras.losses.Loss` instance '
+                       'for loss_fn is required.')
+
+    if not isinstance(optimizer, tf.keras.optimizers.Optimizer):
+      raise ValueError('`tf.keras.optimizers` instance '
+                       'for optimizer is required.')
+
+    build_time_list, avg_epoch_time_list, train_step_time_list = [], [], []
+    wall_time_list, exp_per_sec_list, warmup_time_list = [], [], []
+
+    total_num_examples = epochs * self.num_examples
+
+    for _ in range(run_iters):
+      timer = timeit.default_timer
+      start_time = timer()
+
+      strategy_scope = distribution_util.get_strategy_scope(
+          distribution_strategy)
+      with strategy_scope:
+        t0 = timer()
+        model = model_fn()
+        build_time = timer() - t0
+      # Run one warm up epoch.
+      t1 = timer()
+      self.train_function(
+          model, dataset, loss_fn, optimizer,
+          1, distribution_strategy, batch_size)
+      warmup_time = timer() - t1
+
+      t2 = timer()
+      train_step_time = self.train_function(
+          model, dataset, loss_fn, optimizer,
+          epochs, distribution_strategy, batch_size)
```

Comment:
I think here should be a large epochs, rather than 2, to avoid large variations.

Reply:
Yeah, I can set it default as 10. It should be fine.